models/Chart.js

```diff
@@ -54,6 +41,28 @@ Chart.belongsTo(Chart, {
     foreignKey: 'forked_from'
 });
 
+Chart.prototype.getPublicId = async function() {
+    let useHash = false;
+
+    if (this.id && this.createdAt && chartIdSalt) {
+        useHash = true;
+    } else if (this.organization_id) {
```

Comment:
maybe I'm misreading the code, but doesn't *every* chart have an id and a `createdAt` timestamp? In this case we would never enter the `else` part of the statement. But let's say the chart doesn't have a `createdAt` timestamp for some reason and based on the team setting we decide that `useHash` should be `true`, I'm pretty sure that line 63 will blow up then because it's using `this.createdAt.toISOString()`.

So I think this needs to be thought over again.

Reply:
Yes, I think you're overlooking the third part of the if statement, which checks for `chartIdSalt`, a config option.

I also don't know what exactly the purpose of the first two parts (`this.id && this.createdAt`) is, as I would also assume that they are always present. I simply carried these over from the existing implementation, and didn't want to delete them without understanding when they are relevant. Maybe there are some edge cases when chart objects are created where these aren't set yet?

Reply:
sorry to be pedantic, but the code still has two problems:

1. if for some reason `this.createdAt` is undefined then the `hash.update()` line will throw an exception (can't access method toISOString of undefined). Since the check seems to be there for some reason I assume it's worth keeping them. I think a Chart instance has no createdAt timestamp before it is first saved to the database.

2. if `chartIdSalt` is undefined or empty but the team setting `publishTarget.hash_publishing` is activated, then `useHash` will be true but the hash will be very predictable because it only contains public information (the chart.id + the createdAt timestamp plus some `--` characters). I suggest that if we want to allow overwriting the global hash config we should actually store a team hash salt in `publishTarget.hash_publishing.

Not really a problem, but I wonder why we went for `hash_publishing` instead of `hashPublishing.`

Reply:
> I also don't know what exactly the purpose of the first two parts (this.id && this.createdAt) is, as I would also assume that they are always present.

This check wasn't always there and caused some errors because `this.createdAt` did actually not  exist. Since then I implemented it. I don't know the exact circumstances but Gregors explanation might be part of it.

Reply:
thanks for the suggestions, i implemented the check for `createdAt`, and also require a configurable hash salt now.

> Not really a problem, but I wonder why we went for hash_publishing instead of hashPublishing.

just for 'consistency' with the other config options under `team.settings.publishTarget`, which are all snake case.